### PR TITLE
find as you type search

### PR DESCRIPTION
**Pull request by sethladd**
WIP

@devoncarew @keertip the plan is to generate a simple JSON file, enough to power a find-as-you-type search.

Do either of you have an opinion on how to implement the HTML bits for this? (e.g. found a good example?)

### CODE REVIEW

**Comment by sethladd** [comment]
cc @keertip @devoncarew please take a look. I think this is good enough to try out and send feedback.

File: lib/src/html_generator.dart [hunk 1]
```diff
@@ -166,13 +168,34 @@ class HtmlGeneratorInstance {
 
     if (package != null) {
       _generateDocs();
+      _generateJson();
```
Comment (devoncarew) [inline lib/src/html_generator.dart]: Perhaps `_generateIndex()` or `_generateSearchIndex()`?

File: lib/src/model.dart [hunk 2]
```diff
@@ -48,7 +50,19 @@ void _addToImplementors(Class c) {
   }
 }
 
-abstract class ModelElement implements Comparable {
+/// An element that is enclosed by some other element.
+///
+/// Libraries are not enclosed.
+abstract class EnclosedElement {
```
Comment (devoncarew) [inline lib/src/model.dart]: +1, nice addition to the model

File: lib/resources/styles.css [hunk 3]
```diff
@@ -606,8 +616,81 @@ button {
...
+}
+
+.tt-hint {
+  color: #999
+}
+
+.tt-menu {
+  right:0;
+  left: inherit !important;
+  width: 422px;
```
Comment (devoncarew) [inline lib/resources/styles.css]: It's not clear from the screenshots how this works on mobile. My sense is that smaller then some screen width, the search menu should display as wide as the screen.

**Comment by devoncarew** [comment]
Looks great! A question about mobile behavior.

File: lib/src/html_generator.dart (hunk 1 above)
Comment (sethladd) [inline lib/src/html_generator.dart]: done!

File: lib/resources/styles.css (hunk 3 above)
Comment (sethladd) [inline lib/resources/styles.css]: I haven't taken a look at mobile yet. That will certainly need another batch of fixes/changes.

**Comment by sethladd** [comment]
Ready for review. It's not perfect, but I believe it's enough to get feedback on the actual behavior.

This also fixes a few issues with paddings and lines.

Search box is hidden for very small widths, will need to revisit. Should be visible for nearly all desktop users.

cc @devoncarew

**Comment by devoncarew** [comment]
It's a big one! Awesome to have this feature. :fire:

One bit of feedback on the UI. I'm not sure what the material design spec for search boxes is, but the square box doesn't say search to me. Consider putting a `border-radius: 2px` on it to round the edges.

lgtm

**Comment by sethladd** [comment]
> Consider putting a border-radius: 2px on it to round the edges.

Sure, I'll give it a try.

Thanks for the review!

**Pull request closed without merging**